File: hw1/sort.js
/*
* This module reads an input file, sorts the numbers by a given sorting method and stores the result into an output file.
*
*/

let fs = require('fs');
let mergeSort = require('./merge-sort-func');
let insertSort = require('./insert-sort-func');

module.exports = function (input, output, sortType) {
    //Read the data file
    let numbersArr = parseInputFile(input);
    //For each line of data file, sort the integers and write result into data file
    let result = sort(numbersArr, sortType);
    //Save the result
    fs.writeFileSync(output, result);
};

function parseInputFile(inputFile) {
    let numbers = [];
    let dataFile = fs.readFileSync(inputFile, 'utf8');

    dataFile.split('\n').forEach((line) => {
        let arr = line.split(' ');
        //remove the first integer from the line string because
        // it denotes the number of integers to sort and it not really needed in JS implementation
        arr.shift();

        //Cast strings to numbers
        for (let i = 0; i < arr.length; i++) {
            arr[i] = Number(arr[i]);
        }
        numbers.push(arr);
    });

    return numbers;
}

function sort(numbersArr, sortingMethod) {
    let result = '';

    numbersArr.forEach(arr => {
        switch (sortingMethod) {
            case 'merge':
                arr = mergeSort(arr);
                break;
            case 'insert':
                arr = insertSort(arr);
                break;
        }

        result += arr.join(' ');
        result += '\n';
    });

    return result;
}

File: hw1/python/sort.py
from merge_sort_func import merge_sort
from insert_sort_func import insert_sort

def parse_and_sort(input, output, sort_type):
    numbers = parse_input_file(input)
    result = sort(numbers, sort_type)
    write_to_file(result, output)
    return None


def parse_input_file(input):
    numbers = []
    f = open(input, "r")
    lines = f.readlines()

    for line in lines:
        line = line.strip()
        line_arr = line.split()
        line_arr.pop(0)

        # Convert strings to numbers
        for i in range(len(line_arr)):
            line_arr[i] = float(line_arr[i])

        numbers.append(line_arr)
    return numbers


def sort(numbers, sort_type):
    result = ""

    for arr in numbers:
        if sort_type == "merge":
            sorted_arr = merge_sort(arr)
        elif sort_type == "insert":
            sorted_arr = insert_sort(arr)

        result += ''.join(str(int(num)) + ' ' for num in sorted_arr)
        result += '\n'

    return result


def write_to_file(str, file_name):
    f = open(file_name, "w")
    f.write(str)
    return None

